go-controller/pkg/ovn/logical_switch_cache.go

```diff
@@ -0,0 +1,148 @@
+package ovn
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	ipam "k8s.io/kubernetes/pkg/registry/core/service/ipallocator"
+
+	"reflect"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	"k8s.io/klog"
+)
+
+// logicalSwitchInfo contains information corresponding to the node. It holds the
+// subnet allocations (v4 and v6) as well as the IPAM allocator instances for each
+// subnet managed for this node
+// logicalSwitchInfo is used to save switch specific information such as subnets,
+// allocated IPs per switch etc.
+type logicalSwitchInfo struct {
+	hostSubnets []*net.IPNet
+	ipams       []ipam.Interface
+}
+
+type logicalSwitchCache struct {
+	cache map[string]logicalSwitchInfo
+	// A RW mutex for logicalSwitchCache which holds logicalSwitch information
+	sync.RWMutex
+}
+
+func (cache *logicalSwitchCache) AddToLogicalSwitchCache(nodeName string, hostSubnets []*net.IPNet) error {
+	cache.Lock()
+	defer cache.Unlock()
+	if existing, ok := cache.cache[nodeName]; ok && !reflect.DeepEqual(existing.hostSubnets, hostSubnets) {
+		klog.Warningf("Node %q logical switch already in cache with subnet %s; replacing with %s", nodeName,
+			util.JoinIPNets(existing.hostSubnets, ","), util.JoinIPNets(hostSubnets, ","))
+	}
+	var ipams []ipam.Interface
+	for _, subnet := range hostSubnets {
+		ipam := ipam.NewCIDRRange(subnet)
+		if ipam == nil {
+			klog.Warningf("IPAM for subnet %s was not initialized for node %q", subnet, nodeName)
+			continue
+		}
+		// reserve IPs for gw, management port, hybrid overlay etc.
+		if err := reserveIps(subnet, ipam); err != nil {
+			klog.Warningf("Failed reserving IPs: %s", err)
+			continue
+		}
+		ipams = append(ipams, ipam)
+	}
+	cache.cache[nodeName] = logicalSwitchInfo{hostSubnets: hostSubnets,
+		ipams: ipams}
+	return nil
+}
+
+func (cache *logicalSwitchCache) DeleteFromLogicalSwitchCache(nodeName string) error {
+	cache.Lock()
+	defer cache.Unlock()
+	delete(cache.cache, nodeName)
+	return nil
+}
+
+func (cache *logicalSwitchCache) SwitchExists(nodeName string) bool {
+	cache.RLock()
+	defer cache.RUnlock()
+	exists, ok := cache.cache[nodeName]
+	if ok && exists.hostSubnets != nil && len(exists.hostSubnets) > 0 {
+		return true
+	}
+	return false
+}
+
+func (cache *logicalSwitchCache) SwitchSubnet(nodeName string) *net.IPNet {
```

Comment:
@abhat can this be `GetSwitchSubnet` or `GetSubnetForSwitch` ?

Reply:
Sure will rename.